<a href="https://colab.research.google.com/github/sourabhumarani/hotel_booking_analysis/blob/main/Hotel_Booking_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Have you ever wondered when the best time of year to book a hotel room is? Or the optimal length of stay in order to get the best daily rate? What if you wanted to predict whether or not a hotel was likely to receive a disproportionately high number of special requests? This hotel booking dataset can help you explore those questions!**
# **This data set contains booking information for a city hotel and a resort hotel, and includes information such as when the booking was made, length of stay, the number of adults, children, and/or babies, and the number of available parking spaces, among other things. All personally identifying information has been removed from the data.**
# **Explore and analyze the data to discover important factors that govern the bookings.**

We all are aware that Tourism & Leisure has led to exponential growth for the Hotel Booking Industry.
The Hotel booking Industry is channelled through conventional offline booking to emerging online booking through apps/websites.

Hotel Booking helps the customers to secure their stay in their preferred type of hotel through their preferred agent or app with desired meal/requests in advance.
Although this gives Hotels/Related Hospitality industries advantages like being prepared for catering to the client needs, it can also create some challenges like last minute cancellations which may affect the room allocation process.

The answers to the usual questions like best time of year to book a hotel room is? Or the optimal length of stay in order to get the best daily rate? can be found with finding patterns in booking data of that hotels/hospitality industry, which can help them to prepare better for upcoming bookings & serve their customers better. 

Some key patterns that can help to find these insights are : 

1. Type of Hotel Customer opts the most.
2. Customer's preferable hotel in relation to number of people they are
3. Average Monthly Booking over the years
4. Cancellation of Booking based on different factors (Lead Time, Deposit Type, Hotel Type, etc.)
5. Average Daily Rate by different room type.
6. Type of Meals that are ordered the most.
7. The Number of days that customers stay in different hotel types & in different time of year.
8. Number of Special Requests made over reservations
9. Repeated Guests percentage for each hotel. 
10. Booking influenced over different market segments and distribution channels
And many more.



# **Importing Libraries & Dataset**

**Importing important libraries : 
 Numpy & Pandas for Numerical Operations, Data Analysis & Manipulation
 Seaborn, Plotly & Matplotlib for Graph, Chart Creation & Data Visualization**

In [ ]:
#import important libraries

import numpy as np
import pandas as pd
import seaborn as sns
import plotly.express as px
import matplotlib.pyplot as plt
%matplotlib inline

**Importing the required Dataset**

In [ ]:
#here we have imported the dataset by uploading it to the github rather than mounting the drive everytime to the colab notebook.
dataset_url = 'https://raw.githubusercontent.com/sourabhumarani/hotel_booking_analysis/main/hotel_bookings.csv'
df = pd.read_csv(dataset_url)

# **Data Understanding**

**Exploring the Imported Data Set**

In [ ]:
df

hotel  is_canceled  lead_time  arrival_date_year  \
0       Resort Hotel            0        342               2015   
1       Resort Hotel            0        737               2015   
2       Resort Hotel            0          7               2015   
3       Resort Hotel            0         13               2015   
4       Resort Hotel            0         14               2015   
...              ...          ...        ...                ...   
119385    City Hotel            0         23               2017   
119386    City Hotel            0        102               2017   
119387    City Hotel            0         34               2017   
119388    City Hotel            0        109               2017   
119389    City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
0                    July                        27   
1                    July                        27   
2                    July                        27   
3                    July                        27   
4                    July                        27   
...                   ...                       ...   
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  ...  deposit_type  agent company  \
0                          0       2  ...    No Deposit    NaN     NaN   
1                          0       2  ...    No Deposit    NaN     NaN   
2                          1       1  ...    No Deposit    NaN     NaN   
3                          1       1  ...    No Deposit  304.0     NaN   
4                          2       2  ...    No Deposit  240.0     NaN   
...                      ...     ...  ...           ...    ...     ...   
119385                     5       2  ...    No Deposit  394.0     NaN   
119386                     5       3  ...    No Deposit    9.0     NaN   
119387                     5       2  ...    No Deposit    9.0     NaN   
119388                     5       2  ...    No Deposit   89.0     NaN   
119389                     7       2  ...    No Deposit    9.0     NaN   

       days_in_waiting_list customer_type     adr  \
0                         0     Transient    0.00   
1                         0     Transient    0.00   
2                         0     Transient   75.00   
3                         0     Transient   75.00   
4                         0     Transient   98.00   
...                     ...           ...     ...   
119385                    0     Transient   96.14   
119386                    0     Transient  225.43   
119387                    0     Transient  157.71   
119388                    0     Transient  104.40   
119389                    0     Transient  151.20   

        required_car_parking_spaces  total_of_special_requests  \
0                                 0                          0   
1                                 0                          0   
2                                 0                          0   
3             

In [ ]:
#finding the shape of the dataset

print(df.shape)

(119390, 32)


 **This shows:
Number of Rows in this Data Set : 119390
Number of Columns in this Data Set : 32**

**Get First Five & Last Five row with Head & Tail method**

In [ ]:
df.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  deposit_type  \
0                        0                     0       2  ...    No Deposit   
1                        0                     0       2  ...    No Deposit   
2                        0                     1       1  ...    No Deposit   
3                        0                     1       1  ...    No Deposit   
4                        0                     2       2  ...    No Deposit   

   agent company days_in_waiting_list customer_type   adr  \
0    NaN     NaN                    0     Transient   0.0   
1    NaN     NaN                    0     Transient   0.0   
2    NaN     NaN                    0     Transient  75.0   
3  304.0     NaN                    0     Transient  75.0   
4  240.0     NaN                    0     Transient  98.0   

   required_car_parking_spaces  total_of_special_requests  reservation_status  \
0                            0                          0           Check-Out   
1                            0                          0           Check-Out   
2                            0                          0           Check-Out   
3                            0                          0           Check-Out   
4                            0                          1           Check-Out   

  reservation_status_date  
0              01-07-2015  
1              01-07-2015  
2              02-07-2015  
3              02-07-2015  
4              03-07-2015  

[5 rows x 32 columns]

In [ ]:
df.tail()

hotel  is_canceled  lead_time  arrival_date_year  \
119385  City Hotel            0         23               2017   
119386  City Hotel            0        102               2017   
119387  City Hotel            0         34               2017   
119388  City Hotel            0        109               2017   
119389  City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  ...  deposit_type  agent company  \
119385                     5       2  ...    No Deposit  394.0     NaN   
119386                     5       3  ...    No Deposit    9.0     NaN   
119387                     5       2  ...    No Deposit    9.0     NaN   
119388                     5       2  ...    No Deposit   89.0     NaN   
119389                     7       2  ...    No Deposit    9.0     NaN   

       days_in_waiting_list customer_type     adr  \
119385                    0     Transient   96.14   
119386                    0     Transient  225.43   
119387                    0     Transient  157.71   
119388                    0     Transient  104.40   
119389                    0     Transient  151.20   

        required_car_parking_spaces  total_of_special_requests  \
119385                            0                          0   
119386                            0                          2   
119387                            0                          4   
119388                            0                          0   
119389                            0                          2   

        reservation_status reservation_status_date  
119385           Check-Out              06-09-2017  
119386           Check-Out              07-09-2017  
119387           Check-Out              07-09-2017  
119388           Check-Out              07-09-2017  
119389           Check-Out              07-09-2017  

[5 rows x 32 columns]

**Get list of the number of column labels in the dataset**

In [ ]:
df.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'country', 'market_segment', 'distribution_channel',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type', 'agent',
       'company', 'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests',
       'reservation_status', 'reservation_status_date'],
      dtype='object')

**Lets dive into check unique values in some key columns**

In [ ]:
df['hotel'].unique()

array(['Resort Hotel', 'City Hotel'], dtype=object)

In [ ]:
df['is_canceled'].unique()

array([0, 1])

In [ ]:
df['arrival_date_year'].unique()

array([2015, 2016, 2017])

In [ ]:
df['market_segment'].unique()

array(['Direct', 'Corporate', 'Online TA', 'Offline TA/TO',
       'Complementary', 'Groups', 'Undefined', 'Aviation'], dtype=object)

In [ ]:
df['distribution_channel'].unique()

array(['Direct', 'Corporate', 'TA/TO', 'Undefined', 'GDS'], dtype=object)

In [ ]:
df['arrival_date_month'].unique()

array(['July', 'August', 'September', 'October', 'November', 'December',
       'January', 'February', 'March', 'April', 'May', 'June'],
      dtype=object)

In [ ]:
df['meal'].unique()

array(['BB', 'FB', 'HB', 'SC', 'Undefined'], dtype=object)

In [ ]:
df['is_repeated_guest'].unique()

array([0, 1])

In [ ]:
df['previous_cancellations'].unique()

array([ 0,  1,  2,  3, 26, 25, 14,  4, 24, 19,  5, 21,  6, 13, 11])

In [ ]:
df['reserved_room_type'].unique()

array(['C', 'A', 'D', 'E', 'G', 'F', 'H', 'L', 'P', 'B'], dtype=object)

In [ ]:
df['deposit_type'].unique()

array(['No Deposit', 'Refundable', 'Non Refund'], dtype=object)

In [ ]:
df['customer_type'].unique()

array(['Transient', 'Contract', 'Transient-Party', 'Group'], dtype=object)

In [ ]:
df['reservation_status'].unique()

array(['Check-Out', 'Canceled', 'No-Show'], dtype=object)

 **Lets take a look at what each Column label represents:**

**1.   hotel**

*   Resort Hotel
*   City Hotel

**2. is _canceled**   
 
*   0 as not cancelled booking
*   1 as cancelled booking

**3. lead_time**

*  The time gap between booking and arrival

**4. arrival_date_year**

*   Arrival year of the guest in hotel

**5. arrival_date_month**

*  Arrival month of the guest in hotel

**6. arrival_date_week_number**

*   Arrival week of the guest in hotel

**7. arrival_date_day_of_month**

*   Arrival date of the month of the guest in hotel

**8. stays_in_weekend_nights**

*   Number of weekend nights stayed in hotel

**9. stayed_in_week_nights**

*   Number of week nights stayed in hotel

**10. adults**

*   Number of adults stayed in hotel

**11. children**

*   Number of children stayed in hotel

**12. babies**

*   Number of babies stayed in hotel

**13. meal**

*  Types of meal guests have preferred: BB – Bed & Breakfast, HB – Half board (breakfast and one other meal – usually dinner), FB – Full board (breakfast, lunch and dinner), SC - Self Catering or Undefined.

**14. country**

*  Country from which guests have arrived
 
**15. market_segment**

 * TA - Travel Agents
 * TO - Travel Operstors
 * Others are self explanatory

**16. distribution_channel**
 
 * The channels through which hotel get its booking.
   * TA - Travel Agents
   * TO - Travel Operstors
   * Others are self explanatory

**17. is_repeated_guest**

 * 0 indicates as non repeated guest
 * 1 indicates as repeated guest

**18. previous_cancellation**

*  Number of previous bookings that were cancelled by the customer prior to the current booking.

**19. previous_bookings_not_canceled**

* Number of previous bookings not cancelled by the customer prior to the current booking

**20. reserved_room_type**

* Code for type of room booked prior to arrival

**21. assigned_room_type**

* Code for type of room assigned at arrival

**22. booking_changes**

* Number of time changes made to the booking

**23. deposit_type**

* No Deposit – no deposit was made prior to the booking
* No Refund – deposit was made with no refund condition prior to the booking
* Refundable – deposit was made with refundable condition prior to the booking

**24. agent**

* Code of the travel agenct who made the booking on behalf of guest

**25. company**

* Code of the travel company who made the booking on behalf of guest

**26. day_in_waiting_list**

* Number of days the booking was in the waiting list before it was confirmed to the customer

**27. customer_type**

* Contract - when booking was a part of allotment or any contract.
* Group - when booking was made by a group of people.
* Transient - individuals or groups that stays for short term.
* Transient-party - when the booking is transient but is associated to, at least other transient booking.

**28. adr**

* It is an average daily rate(adr) which states the average rental revenue earned for an occupied room per day.

**29. required_car_parking_spaces**

* Number of car parking spaces required by the customer.

**30. total_of_special_requests**

* Number of special requests made by the customer to the hotel.

**31. reservation_status**

* Canceled – reservation was cancelled 
* Check-Out – checked in and departed  
* No-Show – reservation was made but did not show up

**32. reservation_status_date**

* Day at which last resevation status was made.

**Get the primary information from the data about data types, non-null values in the data, etc. and Descibe the data**

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 30 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

In [ ]:
df.describe()

is_canceled      lead_time  arrival_date_year  \
count  119390.000000  119390.000000      119390.000000   
mean        0.370416     104.011416        2016.156554   
std         0.482918     106.863097           0.707476   
min         0.000000       0.000000        2015.000000   
25%         0.000000      18.000000        2016.000000   
50%         0.000000      69.000000        2016.000000   
75%         1.000000     160.000000        2017.000000   
max         1.000000     737.000000        2017.000000   

       arrival_date_week_number  arrival_date_day_of_month  \
count             119390.000000              119390.000000   
mean                  27.165173                  15.798241   
std                   13.605138                   8.780829   
min                    1.000000                   1.000000   
25%                   16.000000                   8.000000   
50%                   28.000000                  16.000000   
75%                   38.000000                  23.000000   
max                   53.000000                  31.000000   

       stays_in_weekend_nights  stays_in_week_nights         adults  \
count            119390.000000         119390.000000  119390.000000   
mean                  0.927599              2.500302       1.856403   
std                   0.998613              1.908286       0.579261   
min                   0.000000              0.000000       0.000000   
25%                   0.000000              1.000000       2.000000   
50%                   1.000000              2.000000       2.000000   
75%                   2.000000              3.000000       2.000000   
max                  19.000000             50.000000      55.000000   

            children         babies  is_repeated_guest  \
count  119386.000000  119390.000000      119390.000000   
mean        0.103890       0.007949           0.031912   
std         0.398561       0.097436           0.175767   
min         0.000000       0.000000           0.000000   
25%         0.000000       0.000000           0.000000   
50%         0.000000       0.000000           0.000000   
75%         0.000000       0.000000           0.000000   
max        10.000000      10.000000           1.000000   

       previous_cancellations  previous_bookings_not_canceled  \
count           119390.000000                   119390.000000   
mean                 0.087118                        0.137097   
std                  0.844336                        1.497437   
min                  0.000000                        0.000000   
25%                  0.000000                        0.000000   
50%                  0.000000                        0.000000   
75%                  0.000000                        0.000000   
max                 26.000000                       72.000000   

       booking_changes  days_in_waiting_list            adr  \
count    119390.000000         119390.000000  119390.000000   
mean          0.221124              2.321149     101.831122   
std           0.652306             17.594721      50.535790   
min           0.000000              0.000000      -6.380000   
25%           0.000000              0.000000      69.290000   
50%           0.000000              0.000000      94.575000   
75%           0.000000              0.000000     126.000000   
max          21.000000            391.000000    5400.000000   

       required_car_parking_spaces  total_of_special_requests  
count                119390.000000              119390.000000  
mean                      0.062518                   0.571363  
std                       0.245291                   0.792798  
min                       0.000000                   0.000000  
25%                       0.000000                   0.000000  
50%                       0.000000                   0.000000  
75%                       0.000000                   1.000000  
max                       8.000000                   5.000000

# **Data Cleansing**

**Now that we have understood our available dataset lets prepare it for analysis by cleaning unnecessary values/features**

**Lets see how many null values are there in the data by columns**

In [ ]:
df.isnull().sum().sort_values(ascending = False)

country                           488
children                            4
hotel                               0
is_repeated_guest                   0
reservation_status                  0
total_of_special_requests           0
required_car_parking_spaces         0
adr                                 0
customer_type                       0
days_in_waiting_list                0
deposit_type                        0
booking_changes                     0
assigned_room_type                  0
reserved_room_type                  0
previous_bookings_not_canceled      0
previous_cancellations              0
distribution_channel                0
is_canceled                         0
market_segment                      0
meal                                0
babies                              0
adults                              0
stays_in_week_nights                0
stays_in_weekend_nights             0
arrival_date_day_of_month           0
arrival_date_week_number            0
arrival_date

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 30 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            